### PR TITLE
Add ranking matching models

### PR DESCRIPTION
**Pull request by golovenkov**
### Description
Add several ranking matching models for retrieval-based multi-turn response selection:

- Sequential Matching Network (SMN) [Wu et al., 2017]

- Deep Attention Matching Network (DAM) [Zhou et al., 2018]

- Deep Attention Matching Network with Universal Sentence Encoder (DAM-USE-T). This is our new proposed neural ranking architecture based on [Zhou et al., 2018] and [Cer D. et al., 2018].
The model DAM-USE-T outperforms the current state-of-the-art models on the Ubuntu Dialogue Corpus v1 [Lowe et al., 2015] and Ubuntu Dialogue Corpus v2 [Lowe et al., 2017] with the test metrics: `0.7957/0.8932/0.9734` and `0.7446/0.8677/0.9738` accordingly.

**UPD**: added pre-trained models (SMN, DAM, DAM-USE-T) for both Ubuntu Corpus v1 and v2 datasets. There are 6 models in total, see ranking configs, for example, pre-trained model DAM-USE-T for Ubuntu-v1 is  [here](https://github.com/deepmipt/DeepPavlov/blob/57f4bbd869047892eae46e1b29eaf5690d5887e7/deeppavlov/configs/ranking/ranking_ubuntu_v1_mt_word2vec_dam_transformer.json#L123-L126)

### References
Ryan Lowe, Nissan Pow, Iulian Serban, and Joelle Pineau. 2015. The ubuntu dialogue corpus: A large dataset for research in unstructured multi-turn dialogue systems. arXiv preprint arXiv:1506.08909.

Ryan Thomas Lowe, Nissan Pow, Iulian Vlad Serban, Laurent Charlin, Chia-Wei Liu, and Joelle Pineau. 2017. Training end-to-end dialogue systems with the ubuntu dialogue corpus. Dialogue and Discourse 8(1).

Xiangyang Zhou, Lu Li, Daxiang Dong, Yi Liu, Ying Chen, Wayne Xin Zhao, Dianhai Yu and Hua Wu. 2018. Multi-Turn Response Selection for Chatbots with Deep Attention Matching Network. Proceedings of the 56th Annual Meeting of the Association for Computational Linguistics (Volume 1: Long Papers), pages 1118-1127, ACL. http://aclweb.org/anthology/P18-1103 

Yu Wu, Wei Wu, Ming Zhou, and Zhoujun Li. 2017. Sequential match network: A new architecture for multi-turn response selection in retrieval-based chatbots. In ACL, pages 372–381.

Daniel Cer, Yinfei Yang, Sheng-yi Kong, Nan Hua, Nicole Limtiaco, Rhomni St. John, Noah Constant, Mario Guajardo-Cespedes, Steve Yuan, Chris Tar, Brian Strope, Ray Kurzweil. 2018a. Universal Sentence Encoder for English. Proceedings of the 2018 Conference on Empirical Methods in Natural Language Processing: System Demonstrations. Pages 169–174. http://aclweb.org/anthology/D18-2029

### CODE REVIEW

File: deeppavlov/models/preprocessors/ubuntu_v2_preprocessor.py [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+from deeppavlov.core.common.registry import register
+from deeppavlov.core.models.component import Component
+from deeppavlov.core.models.estimator import Estimator
+from deeppavlov.core.data.utils import zero_pad_truncate
+
+log = get_logger(__name__)
+
+
+@register('ubuntu_v2_preprocessor')
+class UbuntuV2Preprocessor(Estimator):
```
Comment (puleon) [inline deeppavlov/models/preprocessors/ubuntu_v2_preprocessor.py]: This class could have more general name as it is used with ubuntu v1 as well as with v2.

File: deeppavlov/models/ranking/deep_attention_matching_network_use_transformer.py [hunk 2]
```diff
@@ -0,0 +1,561 @@
...
+class DAMNetworkUSETransformer(TensorflowBaseMatchingModel):
+    """
+    Tensorflow implementation of Deep Attention Matching Network (DAM) [1] improved with USE [2]
+    ```
+    http://aclweb.org/anthology/P18-1103
+
+    Based on Tensorflow code: https://github.com/baidu/Dialogue/tree/master/DAM
+    We added USE-T [2] as a sentence encoder to the model to achieve state-of-the-art performance on the datasets:
+    * Ubuntu Dialogue Corpus v1 (R@1: 0.739, R@2: 0.867, R@5: 0.972)
+    * Ubuntu Dialogue Corpus v2 (R@1: ?, R@2: ?, R@5: ?)
```
[the remark this answers is not in the document]
Comment (golovenkov) [reply deeppavlov/models/ranking/deep_attention_matching_network_use_transformer.py]: thanks! updated

File: deeppavlov/core/common/registry.json [hunk 3]
```diff
@@ -112,7 +117,10 @@
   "typos_iterator": "deeppavlov.dataset_iterators.typos_iterator:TyposDatasetIterator",
   "typos_kartaslov_reader": "deeppavlov.dataset_readers.typos_reader:TyposKartaslov",
   "typos_wikipedia_reader": "deeppavlov.dataset_readers.typos_reader:TyposWikipedia",
+  "ubuntu_dstc7_mt_reader": "deeppavlov.dataset_readers.ubuntu_dstc7_mt_reader:UbuntuDSTC7MTReader",
```
Comment (puleon) [inline deeppavlov/core/common/registry.json]: Is this reader used in any config?

File: deeppavlov/models/preprocessors/ubuntu_preprocessor.py [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+from deeppavlov.core.common.registry import register
+from deeppavlov.core.models.component import Component
+from deeppavlov.core.models.estimator import Estimator
+from deeppavlov.core.data.utils import zero_pad_truncate
+
+log = get_logger(__name__)
+
+
+@register('ubuntu_preprocessor')
+class UbuntuPreprocessor(Estimator):
```
Comment (puleon) [inline deeppavlov/models/preprocessors/ubuntu_preprocessor.py]: There are very few differences between this class and `SiameseReader` class. Could you please instead add this functionality to the `SiameseReader` using some flags?

File: deeppavlov/models/ranking/siamese_model.py [hunk 5]
```diff
@@ -31,7 +31,7 @@ class SiameseModel(NNModel):
 
     def __init__(self,
                  batch_size: int,
-                 num_context_turns: int = 1,
+                 num_context_turns: int,
```
Comment (puleon) [inline deeppavlov/models/ranking/siamese_model.py]: Why do you remove default value?

File: deeppavlov/models/ranking/siamese_model.py [hunk 6]
```diff
@@ -71,17 +70,22 @@ def __call__(self, batch: Iterable[List[np.ndarray]]) -> Union[np.ndarray, List[
...
-            y_pred = np.reshape(y_pred, (j, len(responses)))
+        # reshape to [batch_size, n_responses] if needed (n_responses > 1)
+        y_pred = np.reshape(y_pred, (j, n_responses)) if n_responses > 1 else y_pred
         return y_pred
 
     def reset(self) -> None:
         pass
 
-    def _train_on_batch(self, batch: List[np.ndarray], y: List[int]) -> float:
+    def _append_sample_to_batch_buffer(self, sample: List, buf: Union[List[List[np.ndarray]], List[Tuple[np.ndarray]]]):
```
Comment (puleon) [inline deeppavlov/models/ranking/siamese_model.py]: Does this function reduce the amount of code?

File: deeppavlov/core/common/registry.json (hunk 3 above)
Comment (golovenkov) [reply deeppavlov/core/common/registry.json]: It is not, actually. But i'm going to use it to score DSTC7 dataset in the future.

File: deeppavlov/models/preprocessors/ubuntu_preprocessor.py (hunk 4 above)
Comment (golovenkov) [reply deeppavlov/models/preprocessors/ubuntu_preprocessor.py]: `SiamesePreprocessor`, you mean?
OK, I will merge the code.

File: deeppavlov/models/ranking/siamese_model.py (hunk 6 above)
Comment (golovenkov) [reply deeppavlov/models/ranking/siamese_model.py]: no changes to `_train_on_batch`:
https://github.com/deepmipt/DeepPavlov/blob/99e7d890b6d366337ac97477bf3d31c048370561/deeppavlov/models/ranking/siamese_model.py#L85

File: deeppavlov/models/ranking/siamese_model.py (hunk 5 above)
Comment (golovenkov) [reply deeppavlov/models/ranking/siamese_model.py]: I think I can revert it back, just need some testing...

File: deeppavlov/models/ranking/siamese_model.py (hunk 6 above)
Comment (puleon) [inline deeppavlov/models/ranking/siamese_model.py]: Does this function reduce the amount of code?

File: deeppavlov/models/ranking/tf_base_matching_model.py [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+        """
+        loss = 0
+        buf = []
+        j = 0
+        while True:
+            try:
+                sample = next(samples_generator)
+                j += 1
+                self._append_sample_to_batch_buffer(sample=sample, buf=buf)
+                if len(buf) >= self.batch_size:
```
Comment (puleon) [inline deeppavlov/models/ranking/tf_base_matching_model.py]: The length of  the `list(samples_generator)` here is always equal to the `self.batch_size`. You could simply convert the samples_generator to the list and use the `_make_batch` function to process it.

File: deeppavlov/models/preprocessors/ubuntu_preprocessor.py (hunk 4 above)
Comment (puleon) [reply deeppavlov/models/preprocessors/ubuntu_preprocessor.py]: Yes, exactly. Thank you!

File: deeppavlov/models/ranking/tf_base_matching_model.py (hunk 7 above)
Comment (golovenkov) [reply deeppavlov/models/ranking/tf_base_matching_model.py]: It still needs to call `self.__append_sample_to_batch_buffer` function, e.g. when dealing with inputs of the SMN/DAM/DAM-USE-T models.
So I write the code that is equal to `list(batch)` but uses `self.__append_sample_to_batch_buffer` to format `buf` as a `List[Tuple[np.ndarray]]`.
Please, see this:
https://github.com/deepmipt/DeepPavlov/blob/ranking_matching_models_use_transformer/deeppavlov/models/ranking/siamese_model.py#L47

File: deeppavlov/models/ranking/siamese_model.py (hunk 5 above)
Comment (golovenkov) [reply deeppavlov/models/ranking/siamese_model.py]: Reverted, breakes nothing. https://github.com/deepmipt/DeepPavlov/blob/8c9e9fd176abbb799b075512d10b1d7effa2ed17/deeppavlov/models/ranking/siamese_model.py#L34

File: deeppavlov/models/preprocessors/ubuntu_preprocessor.py (hunk 4 above)
Comment (golovenkov) [reply deeppavlov/models/preprocessors/ubuntu_preprocessor.py]: Merged.
When `sent_vocab` is not used (i.e. when there is not `sent_vocab` argument in a config), it should not be fit. Is it good to write like:
https://github.com/deepmipt/DeepPavlov/blob/8c9e9fd176abbb799b075512d10b1d7effa2ed17/deeppavlov/models/preprocessors/siamese_preprocessor.py#L103?

File: deeppavlov/core/common/registry.json (hunk 3 above)
Comment (puleon) [reply deeppavlov/core/common/registry.json]: ок.

File: deeppavlov/models/ranking/siamese_model.py (hunk 5 above)
Comment (puleon) [reply deeppavlov/models/ranking/siamese_model.py]: Thank you.

File: deeppavlov/models/ranking/tf_base_matching_model.py (hunk 7 above)
Comment (puleon) [reply deeppavlov/models/ranking/tf_base_matching_model.py]: ok.

File: deeppavlov/models/preprocessors/siamese_preprocessor.py [hunk 8]
```diff
@@ -69,6 +71,7 @@ def __init__(self,
                  use_matrix: bool = True,
                  num_context_turns: int = 1,
                  num_ranking_samples: int = 1,
+                 add_raw_text: bool = False,
                  tokenizer: Component = None,
                  vocab: Estimator = "simple_vocab",
```
Comment (yoptar) [inline deeppavlov/models/preprocessors/siamese_preprocessor.py]:
```suggestion
                 vocab: Optional[Estimator] = None,
```
The same for `embedder`. And then replace
```python3
if type(self.sent_vocab) != str:
```
with 
```python3
if self.sent_vocab is not None:
```

**Comment by golovenkov** [comment]
Just take a while, `MatchingPredictor` and `MatchingPredictorUSE` are useless now, because of removed `*_interact.json` configs for SMN/DAM/DAM-USE-T.
Going to remove them too.

File: deeppavlov/models/ranking/deep_attention_matching_network_use_transformer.py [hunk 9]
```diff
@@ -0,0 +1,532 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import numpy as np
+import tensorflow as tf
+import tensorflow_hub as hub
+from typing import List, Iterable, Dict, Tuple
```
Comment (yoptar) [inline deeppavlov/models/ranking/deep_attention_matching_network_use_transformer.py]: imports order pip8

File: deeppavlov/configs/ranking/ranking_ubuntu_v2_mt_word2vec_dam_transformer.json [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+  },
+  "metadata": {
+    "variables": {
+      "ROOT_PATH": "~/.deeppavlov",
+      "DOWNLOADS_PATH": "{ROOT_PATH}/downloads",
+      "MODELS_PATH": "{ROOT_PATH}/models"
+    },
+    "requirements": [
+      "{DEEPPAVLOV_PATH}/requirements/tf.txt",
+      "{DEEPPAVLOV_PATH}/requirements/gensim.txt"
```
Comment (yoptar) [inline deeppavlov/configs/ranking/ranking_ubuntu_v2_mt_word2vec_dam_transformer.json]: no tf-hub in requirements, but it's needed there.
Check also for other configs

File: tests/test_quick_start.py [hunk 11]
```diff
@@ -136,7 +136,15 @@
...
                  ('IP',)): [ONE_ARGUMENT_INFER_CHECK],
                 ("ranking/paraphrase_ident_qqp_bilstm_test.json", "ranking", ('TI',)): [ONE_ARGUMENT_INFER_CHECK],
-                ("ranking/paraphrase_ident_qqp_interact_test.json", "ranking", ('IP',)): [ONE_ARGUMENT_INFER_CHECK]
+                ("ranking/paraphrase_ident_qqp_interact_test.json", "ranking", ('IP',)): [ONE_ARGUMENT_INFER_CHECK],
+                ("ranking/ranking_ubuntu_v1_mt_word2vec_smn.json", "ranking", ('TI',)): [ONE_ARGUMENT_INFER_CHECK],
+                ("ranking/ranking_ubuntu_v1_mt_word2vec_dam.json", "ranking", ('TI',)): [ONE_ARGUMENT_INFER_CHECK],
+                ("ranking/ranking_ubuntu_v1_mt_word2vec_dam_transformer.json", "ranking", ('TI',)): [ONE_ARGUMENT_INFER_CHECK],
+                ("ranking/ranking_ubuntu_v2_mt_word2vec_smn.json", "ranking", ('TI',)): [ONE_ARGUMENT_INFER_CHECK],
+                ("ranking/ranking_ubuntu_v2_mt_word2vec_dam.json", "ranking", ('TI',)): [ONE_ARGUMENT_INFER_CHECK],
+                ("ranking/ranking_ubuntu_v2_mt_word2vec_dam_transformer.json", "ranking", ('TI',)): [ONE_ARGUMENT_INFER_CHECK],
```
[the remark this answers is not in the document]
Comment (golovenkov) [reply tests/test_quick_start.py]: Thank you for pointing to the issue.
I need to test: 1) ability to train model, then interact model; 2) ability to interact the pretrained model.
For this 2 tests I have 1 config: https://github.com/deepmipt/DeepPavlov/blob/ranking_matching_models_use_transformer/deeppavlov/configs/ranking/ranking_ubuntu_v2_mt_word2vec_dam_transformer.json
In fact that the previous `TI` test `("ranking/ranking_ubuntu_v2_mt_word2vec_dam_transformer.json", "ranking", ('TI',)): [ONE_ARGUMENT_INFER_CHECK],` passes, then the model is able to interact, isn't it?
Also the model is able to interact locally, I have no idea what can be wrong.
Do we have more logs?

[the remark this answers is not in the document]
Comment (seliverstov) [reply tests/test_quick_start.py]: It seems that the model returns an empty response for the test input. Ad `p.expect(">> ")` waits endlessly for appropriate text pattern in output.

[the remark this answers is not in the document]
Comment (seliverstov) [reply tests/test_quick_start.py]: Inspect code of `TestQuickStart::interact` method for more details and try to run
`python -m deeppavlov interact ranking_ubuntu_v2_mt_word2vec_dam_transformer -d` and check output.

[the remark this answers is not in the document]
Comment (golovenkov) [reply tests/test_quick_start.py]: OK, good!
The model returns an empty string if `Dummy text` provided:
```
x::Dummy text
>>
x::& & & & & & & & bonhoeffer  whar drives do you want to mount what &  i have an ext3 usb drive  & look with fdisk -l & hello there & fdisk is all you need
>> [0.9728276 0.1618539 0.9873016]
x::quit

00000090: 6432 7665 635f 6461 6d5f 7472 616e 7366  d2vec_dam_transf
000000a0: 6f72 6d65 722f 6d6f 6465 6c5f 6461 6d2f  ormer/model_dam/
000000b0: 6d6f 6465 6c0a 783a 3a44 756d 6d79 2074  model.x::Dummy t
000000c0: 6578 740a 3e3e 0a78 3a3a 2620 2620 2620  ext.>>.x::& & &
000000d0: 2620 2620 2620 2620 2620 626f 6e68 6f65  & & & & & bonhoe
000000e0: 6666 6572 2020 7768 6172 2064 7269 7665  ffer  whar drive
000000f0: 7320 646f 2079 6f75 2077 616e 7420 746f  s do you want to
00000100: 206d 6f75 6e74 2077 6861 7420 2620 2069   mount what &  i
00000110: 2068 6176 6520 616e 2065 7874 3320 7573   have an ext3 us
00000120: 6220 6472 6976 6520 2026 206c 6f6f 6b20  b drive  & look
00000130: 7769 7468 2066 6469 736b 202d 6c20 2620  with fdisk -l &
00000140: 6865 6c6c 6f20 7468 6572 6520 2620 6664  hello there & fd
00000150: 6973 6b20 6973 2061 6c6c 2079 6f75 206e  isk is all you n
00000160: 6565 640a 3e3e 205b 302e 3937 3238 3237  eed.>> [0.972827
00000170: 3620 302e 3136 3138 3533 3920 302e 3938  6 0.1618539 0.98
00000180: 3733 3031 365d 0a78 3a3a 7175 6974 0a    73016].x::quit.
```
And  `TestQuickStart::interact` waits for `>> ` (note the trailing whitespace), so the test never ends.
I will rewrite the test to provide the actual data.

File: deeppavlov/models/ranking/deep_attention_matching_network_use_transformer.py (hunk 9 above)
Comment (golovenkov) [reply deeppavlov/models/ranking/deep_attention_matching_network_use_transformer.py]: fixed https://github.com/deepmipt/DeepPavlov/blob/57f4bbd869047892eae46e1b29eaf5690d5887e7/deeppavlov/models/ranking/deep_attention_matching_network_use_transformer.py#L15-L18

**Comment by seliverstov** [comment]
Closed due to inactivity. Reopen when will ready to continue.

**Pull request closed without merging**